src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -183,6 +122,131 @@ public GeneratorDriver AddAdditionalTexts(ImmutableArray<AdditionalText> additio
             return FromState(newState);
         }
 
+        public GeneratorDriverRunResult GetRunResult()
+        {
+            var trees = ArrayBuilder<SyntaxTree>.GetInstance();
+            var diagnostics = ArrayBuilder<Diagnostic>.GetInstance();
+            var results = ArrayBuilder<GeneratorRunResult>.GetInstance();
+            for (int i = 0; i < _state.Generators.Length; i++)
+            {
+                var generator = _state.Generators[i];
+                var generatorState = _state.GeneratorStates[i];
+
+                ArrayBuilder<GeneratedSourceResult> sourceResults = ArrayBuilder<GeneratedSourceResult>.GetInstance();
+                for (int j = 0; j < generatorState.SourceTexts.Length; j++)
+                {
+                    var sourceText = generatorState.SourceTexts[j];
+                    var tree = generatorState.Trees[j];
+
+                    sourceResults.Add(new GeneratedSourceResult(tree, sourceText.Text, sourceText.HintName));
+                }
+
+                results.Add(new GeneratorRunResult(generator, sourceResults.ToImmutableAndFree(), generatorState.Diagnostics, generatorState.Exception));
+                trees.AddRange(generatorState.Trees);
+                diagnostics.AddRange(generatorState.Diagnostics);
+            }
+
+            return new GeneratorDriverRunResult(results.ToImmutableAndFree(), trees.ToImmutableAndFree(), diagnostics.ToImmutableAndFree());
+        }
+
+        internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, DiagnosticBag diagnosticsBag, CancellationToken cancellationToken = default)
+        {
+            // with no generators, there is no work to do
+            if (_state.Generators.IsEmpty)
+            {
+                return _state;
+            }
+
+            // run the actual generation
+            var state = StateWithPendingEditsApplied(_state);
+            var stateBuilder = ArrayBuilder<GeneratorState>.GetInstance();
+            int receiverCount = 0;
+
+            for (int i = 0; i < state.Generators.Length; i++)
+            {
+                var generator = state.Generators[i];
+                var generatorState = state.GeneratorStates[i];
+
+                // initialize the generator if needed
+                if (!generatorState.Info.Initialized)
+                {
+                    generatorState = InitializeGenerator(generator, diagnosticsBag, cancellationToken);
```

Comment:
>InitializeGenerator [](start = 37, length = 19)

It looks like `InitializeGenerator()` is only called here. Consider inlining the call. In particular, that would it make it easier to understand where exceptions are expected and handled in this code path.